### PR TITLE
[RLlib] A2/3C policy sub-classing schema.

### PR DESCRIPTION
**Pull request by sven1977**
A2/3C policy sub-classing schema.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by gjoliver** [comment]
Thanks a ton for this PR Sven. You are so fast ... :)
I think we should merge my last PR migrating PPO, IMPALA, and APPO first (https://github.com/ray-project/ray/pull/25117), since it contains a couple of fixes for TF2 eager that is required here.
Sorry about the delay in sending my PR out, I had to debug what's wrong with TF2 eager with my PR, and turns out, I should only cache the action fetches for TF1.

**Comment by sven1977** [comment]
Sounds great!
No worries. Thanks @gjoliver for pushing this very important RLlib design change forward. We can wait merging this one.

File: rllib/agents/a3c/a3c_tf_policy.py [hunk 1]
```diff
@@ -1,178 +1,181 @@
...
     ModelGradients,
+    TensorType,
 )
 
 tf1, tf, tfv = try_import_tf()
 
 
+# We need this builder function because we want to share the same
+# custom logics between TF1 dynamic and TF2 eager policies.
+def get_a3c_tf_policy(base: Type[Union[DynamicTFPolicyV2, EagerTFPolicyV2]]) -> Type[Union[DynamicTFPolicyV2, EagerTFPolicyV2]]:
```
Comment (gjoliver) [inline rllib/agents/a3c/a3c_tf_policy.py:38]: ah, this type annotation is very nice.
do you think we should create a shortcut name in utils.typing, so it reads a bit shorter here?

File: rllib/agents/a3c/a3c_tf_policy.py [hunk 2]
```diff
@@ -1,178 +1,181 @@
...
+        @override(base)
+        def compute_gradients_fn(
+            self, optimizer: LocalOptimizer, loss: TensorType
+        ) -> ModelGradients:
+            return compute_gradients(self, optimizer, loss)
+
+    return A3CTFPolicy
+
+
+A3CDynamicTFPolicy = get_a3c_tf_policy(DynamicTFPolicyV2)
```
Comment (gjoliver) [inline rllib/agents/a3c/a3c_tf_policy.py:171]: let's rename this A3CStaticGraphTFPolicy now.

**Review by gjoliver — COMMENTED** [review]
looks great!! just a few nits and questions.

File: rllib/agents/a3c/a3c_tf_policy.py (hunk 2 above)
Comment (sven1977) [reply rllib/agents/a3c/a3c_tf_policy.py:171]: Ah, yeah, sorry. Didn't care about my own suggestion :D

File: rllib/agents/a3c/a3c_tf_policy.py [hunk 3]
```diff
@@ -1,178 +1,181 @@
...
-    grad_stats_fn=grad_stats,
-    compute_gradients_fn=clip_gradients,
-    postprocess_fn=compute_gae_for_sample_batch,
-    extra_action_out_fn=add_value_function_fetch,
-    before_loss_init=setup_mixins,
-    mixins=[ValueNetworkMixin, LearningRateSchedule, EntropyCoeffSchedule],
+    error=True,
 )
+def postprocess_advantages(*args, **kwargs):
+    pass
```
[the remark this answers is not in the document]
Comment (gjoliver) [reply rllib/agents/a3c/a3c_tf_policy.py:181]: oh I see, it's an Error now. so we don't need the actual logic.
ok, I am all good.

File: rllib/agents/a3c/a3c_tf_policy.py (hunk 1 above)
Comment (sven1977) [reply rllib/agents/a3c/a3c_tf_policy.py:38]: will do!

Comment (sven1977) [reply rllib/agents/a3c/a3c_tf_policy.py:38]: done

**Review by gjoliver — APPROVED** [review]
Thanks, looks awesome!

**Pull request merged**